src/dmd/expressionsem.d

```diff
@@ -5974,6 +6085,32 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             e.e1 = e.e1.expressionSemantic(sc);
         }
         result = e;
+        if (f.vthis2 && !sc.intypeof && sc.func)
+        {
+            if (AggregateDeclaration ad2 = f.isMember2())
+            {
+                Expression te = new ThisExp(e.loc).expressionSemantic(sc);
+                if (te.op != TOK.error)
+                    te = getRightThis(e.loc, sc, ad2, te, f);
+                if (te.op == TOK.error)
+                {
+                    e.error("need `this` of type `%s` to make delegate from function `%s`", ad2.toChars(), f.toChars());
+                    return setError();
+                }
+            }
+            Type tthis2 = Type.tvoidptr.sarrayOf(2);
```

Comment:
this ast manipulation looks complicated enough to go into a documented function

Reply:
done